### ISSUE TITLE
Edge case making a Container receive a subset of the data asked in the fragment

### ISSUE THREAD

**Issue by gre** [facebook/relay#1266]
if you have an object `{ id: "A", subObject { id: "A_sub", name: "foo" } }` and later, the server data the subObject id happen to change (e.g. another user changed it), like to `{ id: "A", subObject { id: "B_sub", name: "bar" } }`.
Then, in some cases, the component won't ask the new fields, Relay thinks the name field was already loaded and only ask pieces it missed from the previous dataset. it is not handling the fact subObject.id has changed, and then you end up with just  `{ id: "A", subObject { id: "B_sub" } }` in props (I guess because B_sub is not in Relay cache) and this even though you asked for subObject.name in the fragment

so you ends up receive half of the data you asked in your fragment, which should be an invariant, right? (the propTypes validation generated by Relay itself are console.error-ing)

Do you want me to try to provide an app reproducing this or is my explanation clear enough?
Thanks

**Comment by NevilleS** [comment]
Relay fundamentally assumes your **nodes never change** unless you do so explicitly via a mutation or a forced fetch. That means that this is, more or less, expected behaviour (there's no special case on the `id` field that causes Relay to expire that node from the store, for example).

If the child node is changing due to some user action, you should capture this via a mutation so Relay knows about it and can refetch accordingly. For your schema, some kind of `FIELDS_CHANGE` mutation using the parent ID of "A" and a fat query like `{ parent { subObject } }` would suffice.

**Comment by gre** [comment]
I'm afraid mutation can't solve everything.
What if someone else have done a mutation that makes these nodes changes between the 2 steps?
That means all container should assume the data eventually can be broken?

Another use-case I have is, scheduling some changes in the data (e.g. consider the usecase "schedule an article submission in X minutes", I still need a way to refresh data not from a mutation)

If what you said is true, it means Relay is not able to have data shared between users?

**Comment by NevilleS** [comment]
Mutations that occur without your knowledge have to be handled via a `forceFetch`. If you believe you fundamentally can't trust the data for a given container, you should refetch it every time using either the `forceFetch` method on the `relay` prop in your container or on the renderer parent for the container.

**Comment by gre** [comment]
ok. But even using forceFetch, there is still the initial bug I'm talking about. My component gets render with a partial tree, not complete like the fragment requested.

The workaround I found is to use the same fragment everywhere, which is not the Relay idea to put the minimal fragment.

**Comment by NevilleS** [comment]
I must be misunderstanding you then. It makes sense that Relay wouldn't attempt to refetch a field it's already gotten, but it sounds like you are describing something different.

> The workaround I found is to use the same fragment everywhere
Not sure what you mean so an example would help

**Comment by gre** [comment]
It's tricky to explain, so I better try to reproduce the bug via an example. It might take some time to build a full example but I'll try to come back with one ASAP :)

basically it's a workaround I found for this exact bug. If I use the same fragment, Relay don't need to fetch the new missing fields, so it used what he thinks to be the latest one (what was cached in subObject of id:"A") . But as soon as there are missing fields, it will fetch **only** these missing fields, but if subObject now have different id, he won't know what are the other fields value. So it doesn't merge the missing fields with anything, like it would normally.

**Comment by NevilleS** [comment]
I could definitely see a scenario where you have something like:
```
fragment on SubObject {
  id
  name
  colour
}
```

During regular operation you decide that this container is rendering subobject with `id = "A_sub"`. The container gets it's data by looking up the node `A_sub` from the store and rendering those fields. Later, if the `id` changes unexpectedly to `B_sub` without using a `forceFetch` or a mutation, I'm not sure how the container would behave. Part of me would expect it to continue rendering the `A_sub` data so you'd see `this.props.subobject.id === "A_sub"`, but if the container did get updated to start looking up `B_sub`, it's entirely possible that you'd see `this.props.subobject.id === "B_sub"` and `this.props.subobject.name` as `null`. In other words, I'm not sure what the behaviour of a container is when the node data changes unexpectedly, but I *can* say that your best bet is to *not* behave unexpectedly 😄

**Comment by gre** [comment]
FYI, I've created a full example just to try to reproduce what I thought was the bug, with what I initially described. It's here: https://github.com/gre/relay-examples/tree/master/cars

But I can't reproduce the bug I experienced at work...
So, i'll close this issue and if someone find something then we can discuss. But maybe it's one of this hard-to-figure-out backend bug on our side.

thanks for your help @NevilleS
